store/job_storer.go

```diff
@@ -0,0 +1,43 @@
+package store
+
+import (
+	"context"
+	models "github.com/ONSdigital/dp-search-reindex-api/models"
+	"github.com/ONSdigital/log.go/log"
+	"time"
+)
+
+type JobStorer struct {
+	JobsMap map[string]models.Job
+}
+
+// CreateJob creates a new Job resource and stores it in the JobsMap
+func (js *JobStorer) CreateJob(ctx context.Context, id string) (models.Job, error) {
+
+	log.Event(ctx, "creating job", log.Data{"id": id})
+	newJob := models.Job{
+		ID:          id,
+		LastUpdated: time.Now().UTC(),
+		Links: &models.JobLinks{
+			Tasks: "http://localhost:12150/jobs/" + id + "/tasks",
+			Self:  "http://localhost:12150/jobs/" + id,
+		},
+		NumberOfTasks:                0,
+		ReindexCompleted:             time.Time{}.UTC(),
+		ReindexFailed:                time.Time{}.UTC(),
+		ReindexStarted:               time.Time{}.UTC(),
+		SearchIndexName:              "Default Search Index Name",
+		State:                        "created",
+		TotalSearchDocuments:         0,
+		TotalInsertedSearchDocuments: 0,
+	}
+	//Only create a new JobsMap if one does not exist already
+	if js.JobsMap == nil {
+		js.JobsMap = make(map[string]models.Job)
+	}
+
+	js.JobsMap[id] = newJob
```

Comment:
Question: What happens if a job with this ID already exists? I don't know the acceptance criteria for this task but I'm assuming we shouldn't be able to create a new job with an existing ID? Depending on the requirements agreed for this task we should probably also capture the desired behaviour with some tests.

Reply:
The id gets generated by the service as a random uuid so in theory it should not try to create a new job with an existing ID. I'm not sure what would happen if it did but then I'm also not sure whether that scenario is something that needs to be handled (given that it's unlikely to happen). Unfortunately my tech lead is on leave this week, my guess would be that I should leave it as is, but what would you suggest?

Reply:
For me personally there is 2 main reasons why we should check the validity of the the ID & adding tests to cover those scenarios:

1 - Although the ID is generated by this service it's generated in another package. By assuming how the ID generation works and that it always gives us a valid unique value we are coupling packages and breaking encapsulation. We are relying on another package working correctly and not changing in future. As a general rule I personally think it's sensible to _never_ trust any input from an external source even if it is within this application. It is more work but I feel it's worth it because it reduces the chance of potential bugs.

2 - A duplicate ID has potentially destructive consequences - we could lose an existing job record if we were to overwrite it, or we could end up in an inconsistent state and I assume we don't want that. The complexity of implementing a check for a duplicate ID is reasonably small. I'd argue a small amount of extra work to be defensive and protect against this scenario is worth it.

As I said I appreciate requires more work but I personally feel it's worth it because it reduces the change of bugs and errors, provides better test coverage and gives us greater confidence in the behaviour of our apps.

Obviously it's up to you if you want to implement these changes or not. As a compromise do you think you could you ask one of the other tech leads for input - I'll go with whatever they decided.

Reply:
For the sake of a good quality app I will go with what you suggest and I think you're right - it would be nice to be confident that the id won't be duplicated. In that case I will add validation and tests as requested.

Reply:
validation and tests now added